[fix][broker] Fix delete namespace fail by a In-flight topic

mattisonchao opened the pull request:
Fixes #19083 
Fixes #11866 

### Motivation

Currently, the force delete namespace operation has the following steps:

1. Do pre-check
2. Get full of topics
3. Mark the namespace as `deleted` to avoid creating the topic again by client lookup or reconnect.
4. Delete all of the user-created topics
5. Delete all system topics.
6. Delete namespace event topics.
7. clean up namespace metadata and resources

A race condition will make step 2 unable to get full of topics. e.g:

| -      | Thread A                                                                                 | Thread B                                                    |
|--------|------------------------------------------------------------------------------------------|-------------------------------------------------------------|
| time 1 | Got full of topics                                                                       | Trying to create a new topic and passed the `deleted` check |
| time 2 | Mark `deleted`                                                                           | Do other checks                                             |
| time 3 | Do the rest of steps 4,5,6                                                               | Created managed ledger and has persistent info to metadata  |
| time 4 | Step 7, Got exception `Directory not empty for /managed-ledgers/test-tenant/test-ns2/persistent` | Do the rest work                                            |
| time 5 | Return the exception                                                                     | Topic created                                               |
 

### Modifications

> Deleting topic xxxx because local cluster is not part of global namespace repl list

- Remove `namespacePolicies.deleted` logic to ensure the new topic will not be deleted by `PersistenTopic#checkReplication`.
- Add retry to resolve some topics that were not created successfully(in metadata) during the deletion.
- Add a test case to test if the topic object is left behind.

### Verifying this change

- [x] Make sure that the change passes the CI checks.


### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

mattisonchao commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:
```diff
@@ -246,9 +246,6 @@ protected void updateTopicPolicyByNamespacePolicy(Policies namespacePolicies) {
         if (log.isDebugEnabled()) {
             log.debug("[{}]updateTopicPolicyByNamespacePolicy,data={}", topic, namespacePolicies);
         }
-        if (namespacePolicies.deleted) {
```
Remove `namespacePolicies.deleted` logic to ensure the new topic will not be deleted by `PersistenTopic#checkReplication`.

eolivelli replied:
with https://github.com/apache/pulsar/pull/19097 we will be sure that "namespacePolicies.deleted" is authoritative.
if we are deleting the namespace then we should do nothing

cc @dlg99 @aymkhalil

mattisonchao replied:
Hi @eolivelli 
I don't understand what you mean.

Will removing this test affect anything?

mattisonchao replied:
@eolivelli 

If I don't remove this check, the creating topic in the deleted namespace will automatically be forcefully deleted by `checkReplication` method because the `replicationClusters` is null. See the logic details:

https://github.com/apache/pulsar/blob/17c58a539315cc4ea39655d4328c5caf55f87d3d/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L1468-L1479

eolivelli replied:
I have checked the code again.
We can keep this change, it is harmless, the topic is deemed to be deleted, we can allow the policy to be updated

mattisonchao commented on pulsar-common/src/main/java/org/apache/pulsar/common/naming/SystemTopicNames.java:
```diff
@@ -81,7 +81,7 @@ public static boolean isTopicPoliciesSystemTopic(String topic) {
             return false;
         }
-        return TopicName.get(topic).getLocalName().equals(NAMESPACE_EVENTS_LOCAL_NAME);
+        return TopicName.get(topic).getLocalName().startsWith(NAMESPACE_EVENTS_LOCAL_NAME);
```
to match partitioned topic name `persistent://public/default/__change_event-partition-0`

eolivelli replied:
This is strange.
it means that sometimes we call this function on the logical name of the partitioned topic and sometimes on a partition.

are we changing the behavior ?

mattisonchao replied:
Ok, I can revert this change.

mattisonchao commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/AdminApi2Test.java:
```diff
@@ -1653,6 +1654,9 @@ public void testDeleteNamespaceWithTopicPolicies() throws Exception {
         admin.topics().createNonPartitionedTopic(namespace + "/tobedeleted");
         // verify namespace can be deleted even without topic policy events
         admin.namespaces().deleteNamespace(namespace, true);
+        final CompletableFuture<Optional<Topic>> eventTopicFuture =
```
Add a test case to test if the topic object is left behind.

eolivelli commented:
This patch is related
https://github.com/apache/pulsar/pull/19097

@dlg99 @aymkhalil PTAL

eolivelli commented:
I am not sure that this holds "Always assume the existence of an event topic "

dlg99 commented:
@mattisonchao can this be done slightly differently:
swap order of steps 2 and 3 (mark deleted first, do the rest, unmark on error).
In combo with https://github.com/apache/pulsar/pull/19097 this should work good enough.

There is still a chance that a topic creation passed the check before ns deletion started, and completed after the ns is marked deleted. I don't think we can solve it easily, this will require either distributed locking or serialization of ns/topic metadata operations e.g. via global system topic.

mattisonchao commented:
@dlg99 

>swap order of steps 2 and 3 (mark deleted first, do the rest, unmark on error).
In combo with https://github.com/apache/pulsar/pull/19097 this should work good enough.

> There is still a chance that a topic creation passed the check before ns deletion started, and completed after the ns is marked deleted. I don't think we can solve it easily, this will require either distributed locking or serialization of ns/topic metadata operations e.g. via global system topic.

As you said. the topic still has the chance to go into the race condition. 

> I don't think we can solve it easily,

We don't need to solve the user-created topic but we should make sure to clean up the system topic.

dlg99 commented:
@mattisonchao what's happening in this PR as I understand:
1. delete all topics + all system topics except for the events one.
2. Then delete the events topic
3. but that brings in some problems with handling of ns.deleted metadata (because of rpc?) so let's remove that logic in some places.

This does not sound like a comprehensive solution (can't topic creation happen after p.2 and still cause namespace deletion to fail?) plus it breaks contract for handling of ns.deleted as @eolivelli moted.

On another note, events topic is being deleted using `admin.topics().deletePartitionedTopicAsync` while for other partitioned topics we use `namespaceResources().getPartitionedTopicResources().deletePartitionedTopicAsync` (in internalDeletePartitionedTopicsAsync). Is it done on purpose?

I'd leave ns.deleted handling logic as it is, set ns.deleted for namespace earlier before actually listing topics (seems to be the right thing to do logically).
Do the rest of the steps with retries/backoff - ns is marked as deleted so topic creations should settle at some point. List topics again and delete again. then delete the namespace metadata.

mattisonchao commented:
>I'd leave ns.deleted handling logic as it is, set ns.deleted for namespace earlier before actually listing topics (seems to be the right thing to do logically).
Yes, that is what I'm changing right now. I would move the `markDelete` before listing topics and handling the duplicate deletion exception.

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:
```diff
@@ -248,9 +245,8 @@ protected CompletableFuture<Void> internalDeleteNamespaceAsync(boolean force) {
-                                partitionedTopicPolicy.add(topic);
-                            } else {
+                            if (!TopicName.get(topic).getLocalName()
+                                    .startsWith(SystemTopicNames.NAMESPACE_EVENTS_LOCAL_NAME)) {
```
can we create an utility method in SystemTopicNames ?

mattisonchao commented:
@eolivelli @dlg99 Please review it again. I change the logic to retry.

dlg99 commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -346,7 +346,7 @@ public CompletableFuture<Void> initialize() {
 
                     Policies policies = optPolicies.get();
 
-                    this.updateTopicPolicyByNamespacePolicy(policies);
```
nit: unnecessary change for this PR

mattisonchao replied:
fixed

dlg99 reviewed (approved):
LGTM assuming you address Enrico's comments + CI passes (currently "You have 1 Checkstyle violation")

mattisonchao commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -346,7 +346,7 @@ public CompletableFuture<Void> initialize() {
                     Policies policies = optPolicies.get();
 
-                    this.updateTopicPolicyByNamespacePolicy(policies);
+                    updateTopicPolicyByNamespacePolicy(policies);
```
```suggestion
                    this.updateTopicPolicyByNamespacePolicy(policies);
```

mattisonchao commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:
```diff
@@ -204,23 +208,48 @@ protected CompletableFuture<List<String>> internalGetNonPersistentTopics(Policie
+        RetryUtil.retryAsynchronously(() -> internalDeleteNamespaceAsync0(force),
+                new BackoffBuilder()
+                        .setInitialTime(200, TimeUnit.MILLISECONDS)
+                        .setMandatoryStop(15, TimeUnit.SECONDS)
```
Hi, @eolivelli 
I am trying to use the number of retrying to instead back off. 

>what happens if the user gives up waiting and then it issues again the same command while the backoff is still running ? Ideally everything should work well and the second execution should wait for the previous execution to complete. The expectation for the user is that when the command completes the namespace is deleted.

For this problem, we can give some kind of the same operation a distributed lock to avoid calling the same operation concurrently. I can send the discussion to the mailing list.

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:
```diff
@@ -317,7 +316,21 @@ protected CompletableFuture<List<String>> internalGetNonPersistentTopics(Policie
+                        final Throwable rc = FutureUtil.unwrapCompletionException(error);
+                        if (rc instanceof MetadataStoreException) {
+                            if (rc.getCause() != null && rc.getCause() instanceof KeeperException.NotEmptyException) {
+                                internalRetryableDeleteNamespaceAsync0(force, retryTimes, callback);
```
please log here that you are going to retry

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:
```diff
@@ -204,23 +206,46 @@ protected CompletableFuture<List<String>> internalGetNonPersistentTopics(Policie
+    private void internalRetryableDeleteNamespaceAsync0(boolean force, int retryTimes,
+                                                        @Nonnull CompletableFuture<Void> callback) {
+        if (retryTimes == 0) {
+            // drop out recursive
```
we should fail here and we should not never reach here

mattisonchao replied:
Yes, I think we can move this check before doing recursive.

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:
```diff
@@ -299,7 +314,23 @@ protected CompletableFuture<Void> internalDeleteNamespaceAsync(boolean force) {
+                    if (error != null) {
+                        final Throwable rc = FutureUtil.unwrapCompletionException(error);
+                        if (rc instanceof MetadataStoreException) {
+                            if (rc.getCause() != null && rc.getCause() instanceof KeeperException.NotEmptyException) {
```
here we must check that the number of retries is exhausted and do not perform the recursion

mattisonchao replied:
Sure, Thanks to you. fixed.

eolivelli reviewed (changes requested):
The retry logic is broker

please check my comments

eolivelli reviewed (approved):
LGTM

eolivelli reviewed (approved):
LGTM
